### PR TITLE
feat(product): add state for composite products

### PR DESCRIPTION
**Pull request by lderrickable**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

- A composite product has a discount if the primary product or any of the applied options of the product items have a discount.
- The price is determined by adding up the primary product price and the applied option prices.
- The default option (from the backend) will be set as the initially applied option for each item of the composite product. If no default is chosen by the backend, daffodil will set the first option as the default if the item is required. If the item is not required, and there is no default option, there will be no option set.

### CODE REVIEW

File: libs/product/src/facades/composite-product/composite-product-facade.interface.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { Observable } from 'rxjs';
+import { Action } from '@ngrx/store';
+
+import { DaffStoreFacade } from '@daffodil/core';
+
+export interface DaffCompositeProductFacadeInterface extends DaffStoreFacade<Action> {
```
Comment (griest024) [inline libs/product/src/facades/composite-product/composite-product-facade.interface.ts:6]: Shouldn't the facade also expose the applied options?

File: libs/product/src/reducers/composite-product-entities/composite-product-entities.reducer.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+			[item.id]: getDefaultOption(item)
+		}), {})
+	}
+
+	function getDefaultOption(item: DaffCompositeProductItem): string {
+		const defaultOptionIndex = item.options.findIndex(option => option.is_default);
+
+		return defaultOptionIndex < 0 && item.required ?
+			item.options[0].id :
+			item.options[defaultOptionIndex].id;
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/product/src/reducers/composite-product-entities/composite-product-entities.reducer.ts]: Fixed and added tests. Thank for catching this. I just forgot about that requirement halfway through writing it.

File: libs/product/src/reducers/composite-product-entities/composite-product-entity.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { Dictionary } from '@ngrx/entity';
+
+export interface DaffCompositeProductEntity {
+	id: string;
+	items: Dictionary<string>;
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/product/src/reducers/composite-product-entities/composite-product-entity.ts]: done

File: libs/product/src/selectors/composite-product/composite-product.selectors.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		selectCompositeProductAppliedOptionsEntitiesState,
+		(products, appliedOptionsEntities, props) => {
+			const product = selectProduct.projector(products, { id: props.id });
+			if(product.type !== DaffProductTypeEnum.Composite) {
+				return undefined;
+			}
+
+			return Math.round(100*
+				(selectCompositeProductPrice.projector(products, appliedOptionsEntities, { id: props.id }) - selectCompositeProductDiscountAmount.projector(products, appliedOptionsEntities, { id: props.id }))
+			)/100;
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/product/src/selectors/composite-product/composite-product.selectors.ts:82]: I'm not sure at all. I was getting a weird failure when I subtracted numbers that I think is just a javascript float thing. I think @damienwebdev has a lot more experience with this whole problem so maybe he can chime in.

File: libs/product/src/facades/composite-product/composite-product-facade.interface.ts (hunk 1 above)
Comment (lderrickable) [reply libs/product/src/facades/composite-product/composite-product-facade.interface.ts:9]: yea, whoops. Added

File: libs/product/src/selectors/product-entities/product-entities.selectors.ts [hunk 5]
```diff
@@ -73,13 +74,24 @@ const createProductEntitiesSelectors = <T extends DaffProduct>(): DaffProductEnt
 		(products, props) => products[props.id]
 	);
 
+	const selectProductDiscountAmount = createSelector(
+		selectProductEntities,
+		(products, props) => {
+			const product = selectProduct.projector(products, { id: props.id });
+
+			return !!(product.discount && product.discount.amount) ?
+				product.discount.amount : 0;
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/product/src/selectors/product-entities/product-entities.selectors.ts]: Oh cool, I didn't know you could do that. changed

File: libs/product/src/selectors/composite-product/composite-product.selectors.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (griest024) [reply libs/product/src/selectors/composite-product/composite-product.selectors.ts:82]: Yeah floating point arithmetic is sketchy at best. We should discuss it with @damienwebdev tomorrow.

[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/product/src/selectors/composite-product/composite-product.selectors.ts:82]: @griest024 I spoke to Damien about this. The plan is to convert all prices to Integers when they come into daffodil (the number of decimals places will be a configurable value). That way, we never need to worry about float calculations. I'm going to open a PR for converting prices to Ints, then I'll come back to this PR.

[the remark this answers is not in the document]
Comment (griest024) [reply libs/product/src/selectors/composite-product/composite-product.selectors.ts:82]: sounds good

[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/product/src/selectors/composite-product/composite-product.selectors.ts]: updated with daff arithmetic functions

File: libs/product/src/selectors/composite-product/composite-product.selectors.spec.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+				+ stubCompositeProduct.items[0].options[1].price 
+				- stubCompositeProduct.discount.amount 
+				- stubCompositeProduct.items[0].options[1].discount.amount
+			});
+			expect(selector).toBeObservable(expected);
+		});
+
+		describe('when the price or discount are long decimal values', () => {
+			
+			it('should return the expected discounted price to two decimals', () => {
```
Comment (griest024) [inline libs/product/src/selectors/composite-product/composite-product.selectors.spec.ts:162]: this spec message is no longer correct

Comment (lderrickable) [reply libs/product/src/selectors/composite-product/composite-product.selectors.spec.ts]: fixed

**Pull request merged**